File: src/top/littlefogcat/demolist/luaj/base/VarargTest.java
package top.littlefogcat.demolist.luaj.base;

import org.luaj.vm2.*;
import org.luaj.vm2.lib.jse.CoerceJavaToLua;
import org.luaj.vm2.lib.jse.JsePlatform;
import top.littlefogcat.demolist.luaj.Lua;
import top.littlefogcat.demolist.luaj.LuaUtils;

public class VarargTest {

    public static class Notification {
        public Intent intent = new Intent();

        @Override
        public String toString() {
            return "Notification{" +
                    "intent=" + intent +
                    '}';
        }

        public LuaValue toLuaValue() {
            return CoerceJavaToLua.coerce(this);
        }
    }

    public static class Intent {
        public void send() {
            System.out.println("callback intent send");
        }
    }

    public static final String PATH = Lua.getPath("base/vararg_test.lua");
    public static final Globals GLOBALS = JsePlatform.standardGlobals();
    private static LuaClosure callback;

    public static void main(String[] args) throws InterruptedException {
        LuaClosure func = (LuaClosure) GLOBALS.loadfile(PATH).call();
        Thread.sleep(1000);
        Notification notification = new Notification();
        callback.invoke(notification.toLuaValue());
    }

    public static void varargFun(String type, Object... args) {
        System.out.println("type: " + type);
        for (int i = 0; i < args.length; i++) {
            Object arg = args[i];
            System.out.print("arg" + i + ": " + arg + "/" + arg.getClass().getSimpleName() + "   ");
        }
        System.out.println();
    }

    public static void setTrigger(int type, LuaClosure action, LuaTable args) {
        System.out.println("setTrigger: type = " + type + ", action = " + action + ", args = " + LuaUtils.luaTableToMap(args));
        callback = action;
    }
}

File: src/top/littlefogcat/demolist/luaj/base/vararg_test.lua
VarargTest = luajava.bindClass("top.littlefogcat.demolist.luaj.base.VarargTest")

VarargTest:varargFun("type_test", { "hello", "world", "!", "" })

local action = function(arg0)
    local n = arg0
    print("n = " .. tostring(n))
    local intent = n.intent
    print("intent = " .. tostring(intent))
    intent:send()
end
local params = {
    title = "微信红包"
}
VarargTest:setTrigger(0, action, params)

function string:startsWith(pattern)
    local index = self:find(pattern)
    return index == 1
end

str = "[2条]非洲野牛: 哦"
print("string = " .. str)
--print(str:startsWith("ello"))
--print(str:startsWith("hell"))
--print(str:find("ello"))
--print(str:find("elloe"))
print(str:find("%[%微%信%红%包%]"))
print(str:find("%[2%条%]%非"))

for i = 1, 10 do
    print(string.sub(str, i, i))
end
str.char()

return function(...)
    for i, v in ipairs({ ... }) do
        print(v .. "/" .. type(v))
    end
end